### PR TITLE
Wrap test logic around rendering the breadcrumb.

### PR DESCRIPTION
**Pull request by kadams54**
This branch also lays the foundation to make it easier to test the other render functions in `app.js` on down the road.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4679/
Test FAILed.

**Comment by kadams54** [comment]
## QA

Fire up the app and make sure the breadcrumb still renders.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4680/
Test PASSed.

File: jujugui/static/gui/src/test/test_app_renderer_extension.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      };
+      renderer.db = {
+        environment: {
+          set: utils.makeStubFunction()
+        }
+      };
+      renderer.changeState = utils.makeStubFunction();
+      renderer.showConnectingMask = utils.makeStubFunction();
+      renderer.switchEnv = utils.makeStubFunction();
+      renderer.createSocketURL = utils.makeStubFunction();
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_app_renderer_extension.js]: I can't imagine a time when we wouldn't want to create stubs for these so could you move all of these stubs into the renderer instance on line 46. Centralize the place where we create one master 'stub' then in the beforeEach of the child suites we can overwrite them if need be.

**Comment by hatched** [comment]
:+1: with trivial

QA :ok:

File: jujugui/static/gui/src/test/test_app_renderer_extension.js (hunk 1 above)
Comment (kadams54) [inline jujugui/static/gui/src/test/test_app_renderer_extension.js]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4698/
Test PASSed.

File: jujugui/static/gui/src/test/test_app_renderer_extension.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    renderer.createSocketURL = utils.makeStubFunction();
+  });
+
+  afterEach(function() {
+    renderer.destroy();
+  });
+
+  describe('_renderBreadcrumb', function() {
+    beforeEach(function() {
+      window.juju.components.HeaderBreadcrumb = {};
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_app_renderer_extension.js]: These are the types of things which make this test suite unstable. Modifications to globals which aren't cleaned up afterwards.

You'll want to store whatever this value was before hand, and then reset it to that value in the afterEach

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4700/
Test PASSed.

File: jujugui/static/gui/src/test/test_app_renderer_extension.js (hunk 2 above)
Comment (kadams54) [inline jujugui/static/gui/src/test/test_app_renderer_extension.js]: Good catch. Fixed.

**Comment by huwshimi** [comment]
:+1: thanks @kadams54

**Comment by kadams54** [comment]
:shipit: Thanks all!

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**